File: samples/detection3d/center_point/center_point.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "common/cmdline.hpp"
#include "common/detection3d.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>(
      "model_prefixs", 'm', "model prefixs of the model suite files", false,
      "[/opt/vastai/vaststreamx/data/models/"
      "pointpillar-int8-max-16000_32_10_3_16000_1_16000-vacc/mod]");
  args.add<std::string>("hw_configs", '\0', "hw-config file of the model suite",
                        false, "[]");
  args.add<std::string>(
      "elf_file", '\0', "elf file path", false,
      "/opt/vastai/vaststream/lib/op/ext_op/pointpillar_ext_op");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<uint32_t>("max_points_num", '\0', "max_points_num to run", false,
                     2000000);
  args.add<std::string>("max_voxel_num", '\0', "model max voxel number", false,
                        "[32000]");
  args.add<std::string>("voxel_size", '\0', "model max voxel number", false,
                        "[0.32,0.32,4.2]");
  args.add<std::string>("coors_range", '\0', "model max voxel number", false,
                        "[-50,-103.6,-0.1,103.6,50,4.1]");
  args.add<std::string>("input_file", '\0', "input file", false, "");
  args.add<uint32_t>("shuffle_enabled", '\0', "shuffle enabled", false, 1);
  args.add<uint32_t>("normalize_enabled", '\0', "normalize enabled", false, 1);
  args.add<std::string>("dataset_root", '\0', "dataset root", false, "");
  args.add<std::string>("dataset_output_folder", '\0',
                        "dataset output folder path", false, "");
  args.parse_check(argc, argv);
  return args;
}

vsx::Tensor ReadBinFile(const std::string& binfile, int max_points_num) {
  std::ifstream ifile(binfile.c_str(), std::ios::binary);
  CHECK(ifile.is_open()) << "Failed to open:" << binfile << std::endl;
  ifile.seekg(0, std::ios::end);
  size_t file_len = ifile.tellg();
  ifile.seekg(0, std::ios::beg);
  size_t read_size = file_len < max_points_num * sizeof(uint16_t) * 4
                         ? file_len
                         : max_points_num * sizeof(uint16_t) * 4;
  char* buffer = new char[read_size];
  ifile.read(buffer, read_size);
  ifile.close();
  size_t fp16_size = read_size / sizeof(uint16_t);
  auto data_manager = std::make_shared<vsx::DataManager>(
      read_size, vsx::Context::CPU(), reinterpret_cast<uint64_t>(buffer),
      [](void* ptr) { delete[] reinterpret_cast<char*>(ptr); });
  vsx::Tensor tensor(vsx::TShape({static_cast<int64_t>(fp16_size)}),
                     data_manager, vsx::TypeFlag::kFloat16);
  return std::move(tensor);
}

uint16_t* LoadDataFP32(std::string filePath, int& size) {
  std::ifstream ifile(filePath.c_str(), std::ios::binary);
  int l = ifile.tellg();
  ifile.seekg(0, std::ios::end);
  int m = ifile.tellg();
  size = static_cast<int>((m - l));
  float* hostSpace = reinterpret_cast<float*>(malloc(size));
  ifile.seekg(0, std::ios_base::beg);
  ifile.read(reinterpret_cast<char*>(hostSpace), size);

  uint16_t* hostSpaceFp16 = reinterpret_cast<uint16_t*>(malloc(size / 2));
  for (int i = 0; i < static_cast<int>(size / 4); i++) {
    hostSpaceFp16[i] = vsx::FloatToHalf(hostSpace[i]);
  }
  free(hostSpace);
  size /= 4;
  ifile.close();
  return hostSpaceFp16;
}

vsx::Tensor ReadBinFileFromFP32(const std::string& binfile,
                                int max_points_num) {
  std::ifstream ifile(binfile.c_str(), std::ios::binary);
  CHECK(ifile.is_open()) << "Failed to open:" << binfile << std::endl;
  ifile.seekg(0, std::ios::end);
  size_t file_len = ifile.tellg();
  ifile.seekg(0, std::ios::beg);
  size_t read_size = file_len < max_points_num * sizeof(uint32_t) * 4
                         ? file_len
                         : max_points_num * sizeof(uint32_t) * 4;
  char* buffer = new char[read_size];
  ifile.read(buffer, read_size);
  ifile.close();
  // size_t fp16_size = read_size / sizeof(uint16_t);
  size_t tensor_size = read_size / sizeof(uint32_t);
  auto data_manager = std::make_shared<vsx::DataManager>(
      read_size, vsx::Context::CPU(), reinterpret_cast<uint64_t>(buffer),
      [](void* ptr) { delete[] reinterpret_cast<char*>(ptr); });
  vsx::Tensor fp32_tensor(vsx::TShape({static_cast<int64_t>(tensor_size)}),
                          data_manager, vsx::TypeFlag::kFloat32);

  vsx::Tensor fp16_tensor =
      vsx::Tensor(vsx::TShape({static_cast<int64_t>(tensor_size)}),
                  vsx::Context::CPU(0), vsx::TypeFlag::kUint16);
  float* fp32 = fp32_tensor.MutableData<float>();
  uint16_t* fp16 = fp16_tensor.MutableData<uint16_t>();
  for (int i = 0; i < tensor_size; i++) {
    fp16[i] = vsx::FloatToHalf(fp32[i]);
  }
  return std::move(fp16_tensor);
}

void WriteBinFile(const std::string& binfile, const vsx::Tensor& tensor) {
  std::ofstream ofile(binfile.c_str(), std::ios::binary);
  CHECK(ofile.is_open()) << "Failed to open:" << binfile << std::endl;
  const char* data = tensor.Data<char>();
  size_t bytes = tensor.GetDataBytes();
  ofile.write(data, bytes);
  ofile.close();
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  const int batch_size = 1;
  auto max_voxel_nums =
      vsx::ParseVecUint(args.get<std::string>("max_voxel_num"));
  auto voxel_size = vsx::ParseVecFloat(args.get<std::string>("voxel_size"));

  auto coors_range = vsx::ParseVecFloat(args.get<std::string>("coors_range"));
  auto model_prefixs =
      vsx::ParseVecString(args.get<std::string>("model_prefixs"));
  auto hw_configs = vsx::ParseVecString(args.get<std::string>("hw_configs"));
  CHECK(max_voxel_nums.size() == model_prefixs.size());
  auto max_points_num = args.get<uint32_t>("max_points_num");
  auto shuffle_enabled = args.get<uint32_t>("shuffle_enabled");
  auto normalize_enabled = args.get<uint32_t>("normalize_enabled");

  std::vector<vsx::PPModelConfig> model_configs;
  for (size_t i = 0; i < max_voxel_nums.size(); i++) {
    vsx::PPModelConfig config;
    config.max_voxel_num = max_voxel_nums[i];
    config.hw_config = hw_configs.size() > i ? hw_configs[i] : "";
    config.model_prefix = model_prefixs[i];
    model_configs.push_back(std::move(config));
  }

  vsx::Detection3D detector3d(
      model_configs, args.get<std::string>("elf_file"), voxel_size, coors_range,
      args.get<uint32_t>("device_id"), max_points_num, shuffle_enabled,
      normalize_enabled, 480, 480, 480, 480);

  if (!args.get<std::string>("input_file").empty()) {
    vsx::Tensor input_tensor;
    std::filesystem::path p(args.get<std::string>("input_file"));

    input_tensor = ReadBinFileFromFP32(args.get<std::string>("input_file"),
                                       max_points_num);

    auto tensors_fp16 = detector3d.Process(input_tensor);
    std::vector<vsx::Tensor> tensors_fp32 =
        vsx::ConvertTensorFromFp16ToFp32(tensors_fp16);
    auto scores = tensors_fp32[0].Data<float>();
    auto labels = tensors_fp32[1].Data<float>();
    auto boxes = tensors_fp32[2].Data<float>();
    for (int i = 0; i < 500; i++) {
      if (scores[i] < 0) break;
      std::cout << "label: " << labels[i] << ", score: " << scores[i]
                << ", box:[ ";
      for (int j = 0; j < 7; j++) {
        std::cout << boxes[i * 7 + j] << " ";
      }
      std::cout << "]\n";
    }
  } else {
    std::vector<std::string> filelist;

    // 使用C++17的filesystem库来遍历目录
    auto dataset_root = args.get<std::string>("dataset_root");
    auto dataset_output_folder = args.get<std::string>("dataset_output_folder");
    for (const auto& entry :
         std::filesystem::directory_iterator(dataset_root)) {
      if (entry.is_regular_file() && entry.path().extension() == ".bin") {
        // 将符合条件的文件路径添加到filelist中
        filelist.push_back(entry.path().string());
      }
    }

    std::sort(filelist.begin(), filelist.end());

    for (int s = 0; s < filelist.size(); s++) {
      auto filename = filelist[s];
      std::cout << filename << std::endl;
      vsx::Tensor input_tensor;

      input_tensor = ReadBinFileFromFP32(filename, max_points_num);
      auto output_tensors = detector3d.Process(input_tensor);
      std::filesystem::path p(filename);

      std::vector<vsx::Tensor> tensors_fp32 =
          vsx::ConvertTensorFromFp16ToFp32(output_tensors);
      std::string npz_file =
          dataset_output_folder + "/" + p.stem().string() + ".npz";
      vsx::SaveTensorMap(npz_file, {{"score", tensors_fp32[0]},
                                    {"label", tensors_fp32[1]},
                                    {"boxes", tensors_fp32[2]}});
    }
  }

  return 0;
}

File: samples/detection3d/center_point/center_point.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys
from tqdm import tqdm
import glob
from collections import OrderedDict


current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../../")
sys.path.append(common_path)

from common.detection3d import Detection3D
from easydict import EasyDict as edict

import numpy as np
import argparse
import ast


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefixs",
        default="[/opt/vastai/vaststreamx/data/models/pointpillar-int8-max-16000_32_10_3_16000_1_16000-vacc/mod]",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_configs",
        default="[]",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--max_voxel_num",
        default="[32000]",
        help="model max voxel number",
    )
    parser.add_argument(
        "--voxel_size",
        default="[0.32,0.32,4.2]",
        help="voxel size",
    )
    parser.add_argument(
        "--coors_range",
        default="[-50,-103.6,-0.1,103.6,50,4.1]",
        help="coors range",
    )
    parser.add_argument(
        "--elf_file",
        default="/opt/vastai/vaststream/lib/op/ext_op/pointpillar_ext_op",
        help="elf file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--max_points_num",
        default=2000000,
        type=int,
        help="max points number per input",
    )
    parser.add_argument(
        "--shuffle_enabled",
        default=1,
        type=int,
        help="shuffle enabled",
    )
    parser.add_argument(
        "--normalize_enabled",
        default=1,
        type=int,
        help="normalize enabled",
    )
    parser.add_argument(
        "--input_file",
        default="",
        help="input file",
    )
    parser.add_argument(
        "--save_npz",
        default=0,
        type=int,
        help="save npz file",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="dataset root",
    )
    parser.add_argument(
        "--dataset_output_folder",
        default="",
        help="dataset output folder path",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    if args.dataset_root != "" and os.path.exists(args.dataset_output_folder) is False:
        os.makedirs(args.dataset_output_folder)
    batch_size = 1
    model_prefixs = args.model_prefixs.strip("[").strip("]").split(",")
    hw_configs = args.hw_configs.strip("[").strip("]").split(",")
    max_voxel_nums = ast.literal_eval(args.max_voxel_num)
    voxel_size = ast.literal_eval(args.voxel_size)
    coors_range = ast.literal_eval(args.coors_range)
    max_points_num = args.max_points_num
    shuffle_enabled = args.shuffle_enabled
    normalize_enabled = args.normalize_enabled
    assert len(model_prefixs) == len(max_voxel_nums)

    modle_configs = [
        edict(
            {
                "max_voxel_num": max_voxel_nums[i],
                "model_prefix": model_prefixs[i],
                "hw_config": hw_configs[i] if len(hw_configs) > i else "",
            }
        )
        for i in range(len(model_prefixs))
    ]

    model = Detection3D(
        model_configs=modle_configs,
        elf_file=args.elf_file,
        voxel_sizes=voxel_size,
        coors_range=coors_range,
        shuffle_enabled=shuffle_enabled,
        normalize_enabled=normalize_enabled,
        device_id=args.device_id,
        max_points_num=max_points_num,
        max_feature_width=864,
        max_feature_height=496,
        actual_feature_width=432,
        actual_feature_height=496,
    )

    if args.input_file != "":
        input = None
        input_fp32 = np.fromfile(args.input_file, dtype=np.float32)
        input = input_fp32.astype(np.float16)
        outputs = model.process(input)
        outputs = [out.astype(np.float32) for out in outputs]
        scores, labels, boxes = outputs
        np.set_printoptions(precision=4)
        for i in range(500):
            if scores[i] < 0:
                break
            print(f"label: {int(labels[i])}, score: {scores[i]:.6f}, box:{boxes[i]}")
    else:
        filelist = [
            f for f in sorted(glob.glob(args.dataset_root + "/*.bin", recursive=True))
        ]
        for file in tqdm(filelist):
            print(file)
            input_fp32 = np.fromfile(file, dtype=np.float32)
            input = input_fp32.astype(np.float16)
            scores, labels, boxes = model.process(input)
            np.savez(
                os.path.join(
                    args.dataset_output_folder,
                    file.split("/")[-1].replace("bin", "npz"),
                ),
                **{
                    "score": np.array(scores),
                    "label": np.array(labels),
                    "boxes": np.array(boxes),
                },
            )
